build.gradle.kts

```diff
@@ -56,6 +58,57 @@ subprojects {
     }
 
     tasks.register<DependencyReportTask>("allDependencies") {}
+
+    tasks.register("applyDependencyRules") {
+        doLast {
+
+            fun dependencyError(error: String) {
+                val message = "DEPENDENCY RULE VIOLATION: $error"
+                if (dependencyAnalysis == "fail") {
+                    throw GradleException(message)
+                } else {
+                    println(message)
+                }
+            }
+
+            val dependencies = configurations["compileClasspath"].resolvedConfiguration.resolvedArtifacts
+
+            dependencies.forEach { artifact ->
+                val dependency = artifact.moduleVersion.id
+
+                if (dependency.group == project.group) {
+                    val pathFromRoot = artifact.file.relativeTo(project.rootDir).path
+                    val pathFromThisModule = artifact.file.relativeTo(project.projectDir).path
+
+                    if (!dependency.name.endsWith("-spi") // modules may only depend on `-spi` modules (exceptions follow)
+                        && !dependency.name.endsWith("spi")
```

Comment:
What is the purpose of this line repeating the previous one without the "-"?

Reply:
Fixed to make clearer and added a comment